### PR TITLE
Enable pprof/debug endpoints by default

### PR DESCRIPTION
**Pull request by cpuguy83**
Makes sure that debug endpoints are always available, which will aid in
debugging demon issues.

Wraps debug endpoints in the middleware chain so the can be blocked by
authz.

Part of #31924 which seeks to add a `debug` subcommand to the CLI.

### CODE REVIEW

**Comment by cpuguy83** [comment]
ping @stevvooe @tonistiigi 
Maybe a sec person as well... @cyli

**Comment by stevvooe** [comment]
I'm not against this, but there are security concerns. Might be nice to add something similar to what we have on containerd, which lets you dump these profiles with a [simple command](https://github.com/containerd/containerd/blob/master/cmd/ctr/pprof.go).

@NathanMcCauley @diogomonica PTAL

**Comment by diogomonica** [comment]
@stevvooe if this is exposed alongside the API, do you think it would add any exposure? Arguably if you're doing an AuthZ plugin to stop a `run a container as root`, you could also stop requests to this endpoint.

**Comment by stevvooe** [comment]
@diogomonica With pprof, one can dump parts of memory. Any `AuthZ` plugins hoping to stop running a container as root would have to stop this case, as well.

If you don't thinks its an issue, then let's go forward with it.

**Comment by cpuguy83** [comment]
ping @diogomonica

**Comment by cyli** [comment]
ping @n4ss as well - if this endpoint does get added, this should probably be added to https://github.com/docker/libentitlement, maybe as `engine.pprof`?

Do we have a document for security-sensitive endpoints for AuthZ plugin implementers?

**Comment by cpuguy83** [comment]
@cyli I don't think we do.

If we are ok with the security implications of this, I would like to move this forward.
The major drawback is that authz plugins need to be updated to prevent users from being able to dump memory... unless of course authz plugins are using a whitelist for these kinds of rejections.

**Comment by n4ss** [comment]
> ping @n4ss as well - if this endpoint does get added, this should probably be added to https://github.com/docker/libentitlement, maybe as engine.pprof?

@cyli sure! can you open an issue in https://github.com/docker/libentitlement please for this so we can track and implement that on the entitlement side?

**Comment by tych0** [comment]
I don't think this is much of a security concern, since access to the docker API is already equivalent to root. You can ptrace the docker daemon from a container run with `docker run --privileged --pid=host` and extract whatever memory contents you want today.

That may change if/when dockerd is multitenant, but for now it seems fine.

**Comment by thaJeztah** [comment]
If there _whould_ be a security concern; should we have an option `--disable-pprof` (or whatever) so that users can opt-out?

**Comment by cpuguy83** [comment]
@thaJeztah I'm not sold that it's worth it to do that. This can be blocked by authz as it is.

**Comment by thaJeztah** [comment]
ok, if there's no objections from the security team; I'm moving this to code review

looks like it needs a rebase

**Comment by cpuguy83** [comment]
Rebased

**Comment by thaJeztah** [comment]
Should we document the endpoints in swagger.yml?

**Comment by cpuguy83** [comment]
We could. I originally left them out because as proposed in the referenced issue, we should not treat the debug endpoints as part of the formal API.

However we could add to the swagger.yml and mention as much in the API description.

**Review by thaJeztah — APPROVED** [review]
LGTM

ping @stevvooe @mlaventure @tonistiigi PTAL

File: api/server/router/debug/debug_routes.go
Comment (stevvooe) [inline api/server/router/debug/debug_routes.go]: Might be able to reduce boilerplate with a helper function:

```go
func frameworkAdaptor(handler http.Handler) APIHandler {
  return func(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
    handler(w, r)
    return nil
  }
}

router.NewGetRoute("/pprof/cmdline", frameworkAdaptor(pprof.Cmdline)),
```

File: api/server/router/debug/debug_routes.go
Comment (stevvooe) [inline api/server/router/debug/debug_routes.go]: Can this be replaced with `expvar.Handler()`?

**Comment by cpuguy83** [comment]
Updated

**Comment by stevvooe** [comment]
LGTM

**Comment by thaJeztah** [comment]
> However we could add to the swagger.yml and mention as much in the API description.

Okay, let's have a look at the documentation separately (possibly we want to document separate from the main API ?)

**Pull request merged**
